src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -685,44 +685,47 @@ static int am_recv_event(struct fi_cq_tagged_entry *wc, MPIR_Request * rreq)
 static int am_read_event(struct fi_cq_tagged_entry *wc, MPIR_Request * dont_use_me)
 {
     int mpi_errno = MPI_SUCCESS;
-    MPIR_Request *rreq;
-    MPIDI_OFI_am_request_t *ofi_req;
-
     MPIR_FUNC_ENTER;
 
-    ofi_req = MPL_container_of(wc->op_context, MPIDI_OFI_am_request_t, context);
-    rreq = (MPIR_Request *) ofi_req->rreq_hdr->rreq_ptr;
+    MPIDI_OFI_am_read_req_t *read_req;
+    read_req = MPL_container_of(wc->op_context, MPIDI_OFI_am_read_req_t, context);
+
+    MPIR_Request *rreq;
+    rreq = (MPIR_Request *) read_req->rreq_ptr;
+
 
-    if (ofi_req->rreq_hdr->lmt_type == MPIDI_OFI_AM_LMT_IOV) {
-        ofi_req->rreq_hdr->lmt_u.lmt_cntr--;
-        if (ofi_req->rreq_hdr->lmt_u.lmt_cntr) {
+    if (MPIDI_OFI_AMREQUEST(rreq, rdma_read_hdr)->lmt_type == MPIDI_OFI_AM_LMT_IOV) {
+        if (read_req->lmt_cntr > 0) {
```

Comment:
won't this cause a race? from the looks of it, we always issue all the reads at once. if so, this counter could be zero before all the data has arrived.

Reply:
Note, we are creating separate `read_req` for each read operation, so the counter is local to the particular read operation (, which is set at issuing time). There is no race.